merkledag/merkledag.go

```diff
@@ -112,39 +112,48 @@ func (n *dagService) Remove(nd *Node) error {
 
 // FetchGraph asynchronously fetches all nodes that are children of the given
 // node, and returns a channel that may be waited upon for the fetch to complete
-func FetchGraph(ctx context.Context, root *Node, serv DAGService) chan struct{} {
-	log.Warning("Untested.")
-	var wg sync.WaitGroup
-	done := make(chan struct{})
+func FetchGraph(ctx context.Context, root *Node, serv DAGService) chan error {
+	//done := make(chan error, len(root.Links))
 
+	var keys []key.Key
 	for _, l := range root.Links {
-		wg.Add(1)
-		go func(lnk *Link) {
+		keys = append(keys, key.Key(l.Hash))
+	}
+
+	out := make(chan error)
+	nodes := serv.GetNodes(ctx, keys)
 
-			// Signal child is done on way out
+	var wg sync.WaitGroup
+	for _, n := range nodes {
+		wg.Add(1)
+		go func(n NodeGetter) {
 			defer wg.Done()
-			select {
-			case <-ctx.Done():
-				return
+			nd, err := n.Get(ctx)
+			if err != nil {
+				select {
+				case out <- err:
+					return
+				case <-ctx.Done():
+				}
 			}
 
-			nd, err := lnk.GetNode(ctx, serv)
+			err = <-FetchGraph(ctx, nd, serv)
```

Comment:
this way of doing it spawns a goroutine per node and keeps them all alive until they _all_ get in. maybe a concurrent approach with channels may fare better.